File: 3849-equal-sum-grid-partition-i/3849-equal-sum-grid-partition-i.java
class Solution {
    public boolean canPartitionGrid(int[][] grid) {
        int n = grid.length, m = grid[0].length;
        var rowSum = new long[n];
        var colSum = new long[m];
        long tot = 0;

        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                tot += grid[i][j];
                rowSum[i] += grid[i][j];
                colSum[j] += grid[i][j];
            }
        }

        if ((tot & 1) != 0)
            return false;
        
        long pref = 0L;

        for (int i = 0; i < n - 1; ++i) {
            pref += rowSum[i];
            if ((pref << 1) == tot)
                return true;
        }

        pref = 0L;
        for (int i = 0; i < m - 1; ++i) {
            pref += colSum[i];
            if ((pref << 1) == tot)
                return true;
        }
        
        return false;
    }
}

File: 3849-equal-sum-grid-partition-i/3849-equal-sum-grid-partition-i.py
class Solution(object):
    def canPartitionGrid(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: bool
        """
        n, m = len(grid), len(grid[0])
        rowSum, colSum = [0 for _ in range(n)], [0 for _ in range(m)]
        tot = 0
        for i in range(n):
            for j in range(m):
                tot += grid[i][j]
                rowSum[i] += grid[i][j]
                colSum[j] += grid[i][j]

        if tot & 1:
            return False
            
        pref = 0   
        for i in range(n - 1):
            pref += rowSum[i]
            if (pref << 1) == tot:
                return True

        pref = 0
        for j in range(m - 1):
            pref += colSum[j]
            if (pref << 1) == tot:
                return True
            
        return False